openmc/model/pin.py

```diff
@@ -0,0 +1,249 @@
+        all :class:`openmc.ZCylinders`
+    materials: iterable of :class:`openmc.Material`
+        Materials to go between ``surfaces``. There must be one
+        more material than surfaces, corresponding to the material
+        that spans all space outside the final ring.
+    universe_id: None or int
+        Unique identifier for this universe
+    name: str
+        Name for this universe
+
+    See Also
+    --------
+
+    :meth:`openmc.Pin.from_radii` - Convinience function to build
+    from radii of cylinders
+    """
+
+    def __init__(self, surfaces, materials, universe_id=None, name=""):
+        cv.check_iterable_type("materials", materials, openmc.Material)
+        cv.check_length("surfaces", surfaces, len(materials) - 1)
+
+        # Ensure that all surfaces are same type of cylinder
+        self._check_surfaces(surfaces)
+        regions = subdivide(surfaces)
+
+        cells = [
+            openmc.Cell(fill=m, region=r) for m, r in zip(materials, regions)
+        ]
+
+        super().__init__(universe_id=universe_id, name=name, cells=cells)
+        self._list_cells = cells  # need ordered by radial position
+
+    def _check_surfaces(self, surfaces):
+        cv.check_type(
+            "surface 0",
+            surfaces[0],
+            (openmc.ZCylinder, openmc.YCylinder, openmc.XCylinder),
+        )
+        if isinstance(surfaces[0], openmc.ZCylinder):
+            center_getter = attrgetter("x0", "y0")
+        elif isinstance(surfaces[0], openmc.YClylinder):
+            center_getter = attrgetter("x0", "y0")
+        elif isinstance(surfaces[0], openmc.XClylinder):
+            center_getter = attrgetter("z0", "y0")
+        else:
+            raise TypeError(
+                "Not configured to interpret {} surfaces".format(
+                    surfaces[0].__class__.__name__
+                )
+            )
+        cv.check_iterable_type("surfaces", surfaces[1:], type(surfaces[0]))
+        # Check for concentric-ness and increasing radii
+        centers = set()
+        radii = []
+        rad = 0
+        for ix, surf in enumerate(surfaces):
+            cur_rad = surf.r
+            if cur_rad <= rad:
+                raise ValueError(
+                    "Surfaces do not appear to be increasing in radius. "
+                    "Surface {} at index {} has radius {:7.3E} compared to "
+                    "previous radius of {:7.5E}".format(
+                        surf.id, ix, cur_rad, rad
+                    )
+                )
+            rad = cur_rad
+            radii.append(cur_rad)
+            centers.add(center_getter(surf))
+
+        if len(centers) > 1:
+            raise ValueError(
+                "Surfaces do not appear to be concentric. The following "
+                "centers were found: {}".format(centers)
+            )
+        self._radii = radii
+        self._surfaces = surfaces
+        self._surf_type = type(surfaces[0])
+
+    @classmethod
+    def from_radii(
```

Comment:
Agree. One way this could be useful is if the user had better access to the surfaces on this pin, and wanted to recreate a similar pin, but with different materials. Would hate to define redundant surfaces.

But if they made the surfaces with that list comprehension, they have access to that already